e2e/testcases/reconciler_finalizer_test.go

```diff
@@ -0,0 +1,702 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/api/configsync"
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+	"kpt.dev/configsync/pkg/core"
+	"kpt.dev/configsync/pkg/kinds"
+	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager"
+	"kpt.dev/configsync/pkg/status"
+	"kpt.dev/configsync/pkg/testing/fake"
+	"sigs.k8s.io/cli-utils/pkg/common"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TestReconcilerFinalizer_Orphan tests that the reconciler's finalizer
+// correctly handles Orphan deletion propegation.
+func TestReconcilerFinalizer_Orphan(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.MultiRepos, ntopts.SkipMonoRepo)
+
+	rootRepo := nt.RootRepos[configsync.RootSyncName]
+	deployment1Name := "helloworld-1"
+
+	nt.T.Cleanup(func() {
+		// Stop webhook to avoid deletion prevention.
+		// Webhook will be re-enabled by test setup, if the next test needs it.
+		nomostest.StopWebhook(nt)
+
+		// Delete RootSync
+		rootSync := fake.RootSyncObjectV1Beta1(configsync.RootSyncName)
+		err := nt.Delete(rootSync, client.PropagationPolicy(metav1.DeletePropagationForeground))
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				nt.T.Error(err)
+			}
+		}
```

Comment:
Debatable. It mostly depends on how good the e2e test reset code is, and it's demonstrably broken in some cases. I don't trust it, and I left the rewrite for a follow PR. So I opted for simply deleting everything in reverse order to ensure everything was really gone, and if it wasn't, we got an error explaining why for each object.

Also, the 2nd block will execute first, because the Cleanups are LIFO. But the 2nd cleanup may never happen if it errors before then. The "Orphan" value and deleted annotation _should_ behave the same, but only if everything is working correctly. So I wanted to make sure the annotation was removed (most likely to trigger the existing default behavior) before deleting the RootSync.

Reply:
I moved the annotation removal into the same extracted cleanup function. I think it's best to remove the annotation before deleting it. It's possible we don't actually need to delete the RootSync, but if we don't, I'll have to reset the Git repo to empty out the added objects. Deleting the RootSync seemed safer. That way the test reset can recreate it after it's reset the git repo.